pkg/middleware/stored_session.go

```diff
@@ -105,51 +106,45 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 	return session, nil
 }
 
-// refreshSessionIfNeeded will attempt to refresh a session if the session
-// is older than the refresh period.
+// refreshSessionIfNeeded will attempt to refresh a session
 // It is assumed that if the provider refreshes the session, the session is now
 // valid.
 // If the session requires refreshing but the provider does not refresh it,
 // we must validate the session to ensure that the returned session is still
 // valid.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
-	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
-		// Refresh is disabled or the session is not old enough, do nothing
+	if session == nil {
+		// Refresh is disabled or there is no session, do nothing
 		return nil
 	}
 
-	logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, s.refreshPeriod)
-	refreshed, err := s.refreshSessionWithProvider(rw, req, session)
-	if err != nil {
-		return err
-	}
-
-	if !refreshed {
-		// Session wasn't refreshed, so make sure it's still valid
+	if session.RefreshToken == "" || !s.refresh.ShouldRefresh(time.Now(), *session.CreatedAt, *session.ExpiresOn) {
+		// Session cannot be refreshed due to absent refresh token or has not yet expired
+		// Validate existing session
 		return s.validateSession(req.Context(), session)
 	}
-	return nil
+
+	logger.Printf("Refreshing %s old session cookie for %s", session.Age(), session)
+	return s.refreshSessionWithProvider(rw, req, session)
 }
 
 // refreshSessionWithProvider attempts to refresh the sessinon with the provider
 // and will save the session if it was updated.
-func (s *storedSessionLoader) refreshSessionWithProvider(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) (bool, error) {
-	refreshed, err := s.refreshSessionWithProviderIfNeeded(req.Context(), session)
+func (s *storedSessionLoader) refreshSessionWithProvider(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
+	err := s.refreshSession(req.Context(), session)
 	if err != nil {
-		return false, fmt.Errorf("error refreshing access token: %v", err)
+		return fmt.Errorf("error refreshing access token: %v", err)
```

Comment:
```suggestion
		return fmt.Errorf("error refreshing session tokens: %v", err)
```

NIT: I lean toward making this more generic since some provider implementations will refresh both the ID & Access Tokens.

Reply:
:+1: